iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp

```diff
@@ -30,24 +33,44 @@ namespace roudi
 class ServiceRegistry
 {
   public:
-    static constexpr uint32_t MAX_INSTANCES_PER_SERVICE = 100u;
-    using CaproIdString_t = capro::IdString_t;
-    using InstanceSet_t = cxx::vector<CaproIdString_t, MAX_INSTANCES_PER_SERVICE>;
-    struct instance_t
+    enum class ServiceRegistryError
     {
-        InstanceSet_t instanceSet;
+        INVALID_STATE,
+        SERVICE_DESCRIPTION_ALREADY_ADDED,
+        SERVICE_REGISTRY_FULL,
     };
-    using serviceMap_t = std::map<CaproIdString_t, instance_t>;
+    /// @todo #415 should be connected with iox::MAX_NUMBER_OF_SERVICES
+    static constexpr uint32_t MAX_SERVICE_DESCRIPTIONS = 100U;
+    using CaproIdString_t = capro::IdString_t;
+    using ServiceDescriptionVector_t = cxx::vector<capro::ServiceDescription, MAX_SERVICE_DESCRIPTIONS>;
+
+    /// @brief Adds given service description to registry
+    /// @param[in] serviceDescription, service to be added
+    /// @return ServiceRegistryError, error wrapped in cxx::expected
+    cxx::expected<ServiceRegistryError> add(const capro::ServiceDescription& serviceDescription);
+
+    /// @brief Removes given service description from registry
+    /// @param[in] serviceDescription, service to be removed
+    /// @return true, if service description was removed, false otherwise
+    bool remove(const capro::ServiceDescription& serviceDescription);
```

Comment:
I think it is intended to indicate whether the service was there and removed or was not there and not removed. The logical state of the registry is the same in both cases, so we do not really need to communicate this. Personally I do those kind of things as well for more information if the user wants (the information was already computed and passing it is inexpensive). The user is technically free to ignore the return value (I know there are differing opinions on this one).

This kind of style is definitely used in several APIs, but I am fine either way.

Reply:
> I think it is intended to indicate whether the service was there and removed or was not there and not removed. 

Exactly, that's the idea. There shall be the possibility to check whether remove was successful or remove was unsuccessful, when the element was not contained in the registry.

> And when the services was not added it is still removed.

That's not entirely correct. It will not be removed if it wasn't added before. The `bool` is used to signal this. I'll keep the additional info to be able to pass this to the user via the new `ServiceDiscovery` API.

Reply:
@mossmaurice I disagree. The reason is, what can a user do with this kind of information: One removes a service and after the call it is not there - this is the contract. If you return false I would assume that you tried to remove the service and failed therefore it is still present.

In my opinion it does not make sense to return here a bool. Do you have an argument for that? And if so, why not return an enum so that it is more readable and one really knows what is going on!
Please also consider that we should try to reduce unnecessary code bloat and error handling. If we do return here something we are forced by our standards to handle it and we turn this line
```cpp
serviceRegistry.remove(myService);
```
into
```cpp
if ( serviceRegistry.remove(myService) ) {
   //...
} else {
  //...
}
```

And as final argument: YAGNI, you aint gonna need it. I do not see the use case and if the use case arises we can still add it.